### ISSUE TITLE
[Datasets] Introduce padded batching to allow varying size input to ML models

### ISSUE THREAD

**Issue by jharaldson** [ray-project/ray#29654]
### Description

Introduce padded batching to allow batches of varying sizes as input to ML models. Today we have to do the padding as we write the data to disc. This has several disadvantages; 

1. we have to pad all input according to the largest anticipated input size, resulting in in-efficient data storage.
2. we cannot easily modify the padding type (instead of zeros we might want another strategy) or size of stored data. 

The proposed feature allows us to decide the padding at run time instead of at the time of writing to file. 

In Tensorflow there is a batch operation called [padded_batch](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#padded_batch). 

The padded_batch operation is further explained [here](https://www.tensorflow.org/guide/data#batching_tensors_with_padding).
_"Many models (including sequence models) work with input data that can have varying size (for example, sequences of different lengths). To handle this case, the [Dataset.padded_batch](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#padded_batch) transformation enables you to batch tensors of different shapes by specifying one or more dimensions in which they may be padded."_

**A note**: Not sure how this works with parquet files and data schemas.

### Use case

A clear use case is NLP, but also various data and models related to time series.

The following Tensorflow [guide](https://www.tensorflow.org/guide/keras/masking_and_padding) explains the concept. In essence we want to use padding to get the same size of input for each batch and then we want to mask the loss for the padded parts that should not be taken into consideration.

**Comment by sjkoelle** [comment]
I think we would be very interested in this feature
